File: 21-merge-two-sorted-lists/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1: return list2
        if not list2: return list1

        p,q = list1, list2
        res = None
        if p.val < q.val:
            res = p
            p = p.next
        else:
            res = q
            q = q.next
        r = res
        while p or q:
            if p and q:
                if p.val < q.val:
                    r.next = p
                    p = p.next
                else:
                    r.next = q
                    q = q.next
                r = r.next
            elif p and not q:
                r.next = p
                return res
            else:
                r.next = q
                return res
        return res

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode p1 = list1, p2 = list2, p3=null;
        ListNode newHead = null;

        while(null!=p1 || null!=p2){
            ListNode temp = null;
            // System.out.println(String.format("p1 = %s, p2 = %s, p3 = %s",print(p1), print(p2), print(p3)));
            if(null!=p1 && null!=p2){
                if(p1.val <= p2.val){
                    temp = p1;
                    p1 = p1.next;
                }else{
                    temp = p2;
                    p2 = p2.next;
                }
            }else if(p1==null){
                temp = p2;
                p2 = p2.next;
            }else{
                temp = p1;
                p1 = p1.next;
            }

            // System.out.println("\t temp = "+print(temp));

            if(null==newHead){
                newHead = new ListNode(temp.val);
                p3 = newHead;
            }else{
                p3.next = new ListNode(temp.val);
                p3 = p3.next;
            }
        }

        return newHead;
    }

    public String print(ListNode node){
        return (null==node)?null:""+node.val;
    }
}